services/spar/test-integration/Test/LoggingSpec.hs

```diff
@@ -18,3 +26,33 @@ spec = describe "logging" $ do
         Log.flush logger
       out `shouldContain` "hrgh  woaa"
       out `shouldNotContain` "hrgh\n\nwoaa"
+
+  let withApp :: (Application -> Env -> Expectation) -> TestSpar ()
```

Comment:
This doesn't appear to depend on anything within this test; maybe pull it out into a top-level thing and make it re-usable that way? or otherwise just move it down to a `where` clause?

Reply:
On second thought; maybe just write `buildApp :: TestSpar (Application, Env)` then we can avoid the confusing `continuation passing` style 😄

Reply:
> This doesn't appear to depend on anything within this test; maybe pull it out into a top-level thing and make it re-usable that way? or otherwise just move it down to a `where` clause?

I want to have it hear because I feel strongly that it's where it's easiest to find when I look for it.

Reply:
> This doesn't appear to depend on anything within this test; maybe pull it out into a top-level thing and make it re-usable that way? or otherwise just move it down to a `where` clause?

This won't work: `TestSpar _` is a `description`, but `Expection` is the body of an `it`.

I appreciate most of your feedback (and sorry again for giving you so much to complain about), but I think here things are getting out of hand and we're wasting time.

I will just inline this function away, this will look less tidy because of extra `liftIO`s, but hopefully it's less controversial.

Reply:
> I feel strongly that it's where it's easiest to find when I look for it.

I guess we all have our own subjective preferences. How can we get to a common code style standard though? `withApp` is used in the test below, but also in the ` it "400 on finalize-login...` further down. To me (who's less familiar with this codebase), the current location strikes me as an odd place to put this (reusable) function - I would have expected it elsewhere. What exactly makes this easiest to find for you?

Reply:
yes, i guess the fact that you both disagree is proof enough that it is a question of taste.

since the code has changed and the question does not apply any more, we should probably resolve this thread.  but to answer your question i'll try to reason it out: tests have code blocks often many pages of screen long.  having a where clause at the very bottom that defines local variables to various tests seems like deliberately hiding things to me.  instead, i want to define helpers as locally as possible, sometimes adding an extra do block just so the variable stays local, but certainly on top of where it's used, so that i can read the file from top to bottom, and have everything i need in my very tight short-term memory.

making helpers more general / global when the need arises is trivial.